actors/evm/src/interpreter/precompiles/bls12_381/pairing.rs

```diff
@@ -0,0 +1,432 @@
+};
+use fil_actors_runtime::runtime::Runtime;
+
+use crate::interpreter::precompiles::bls_util::{
+    PADDED_G1_LENGTH, PADDED_G2_LENGTH, PAIRING_INPUT_LENGTH, extract_g1_input, extract_g2_input,
+};
+
+use blst::{
+    blst_final_exp, blst_fp12, blst_fp12_is_one, blst_fp12_mul, blst_miller_loop, blst_p1_affine,
+    blst_p2_affine,
+};
+
+/// BLS12_PAIRING precompile
+/// Implements BLS12-381 pairing check according to EIP-2537
+pub fn bls12_pairing<RT: Runtime>(
+    _: &mut System<RT>,
+    input: &[u8],
+    _: PrecompileContext,
+) -> PrecompileResult {
+    let input_len = input.len();
+
+    // Input length must be non-zero and multiple of the pairing input length
+    if input_len == 0 || input_len % PAIRING_INPUT_LENGTH != 0 {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    // Calculate number of pairs
+    let k = input_len / PAIRING_INPUT_LENGTH;
+
+    // Collect pairs of points for the pairing check
+    let mut pairs = Vec::with_capacity(k);
+
+    for i in 0..k {
+        let encoded_g1_element =
+            &input[i * PAIRING_INPUT_LENGTH..i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH];
+        let encoded_g2_element = &input[i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH
+            ..i * PAIRING_INPUT_LENGTH + PADDED_G1_LENGTH + PADDED_G2_LENGTH];
+
+        // If either the G1 or G2 element is the encoded representation
+        // of the point at infinity, then these two points are no-ops
+        // in the pairing computation.
+        //
+        // Note: we do not skip the validation of these two elements even if
+        // one of them is the point at infinity because we could have G1 be
+        // the point at infinity and G2 be an invalid element or vice versa.
+        // In that case, the precompile should error because one of the elements
+        // was invalid.
+        let g1_is_zero = encoded_g1_element.iter().all(|i| *i == 0);
+        let g2_is_zero = encoded_g2_element.iter().all(|i| *i == 0);
+
+        // NB: Scalar multiplications, MSMs and pairings MUST perform a subgroup check.
+        // extract_g1_input and extract_g2_input perform the necessary checks
+        let p1_aff = extract_g1_input(encoded_g1_element, true)?;
+        let p2_aff = extract_g2_input(encoded_g2_element, true)?;
+
+        if !g1_is_zero & !g2_is_zero {
+            pairs.push((p1_aff, p2_aff));
+        }
+    }
+
+    // Perform the pairing check
+    let result = pairing_check(&pairs);
+
+    // Return 32 bytes: 31 zero bytes followed by 1 for success, 0 for failure
+    let mut output = vec![0u8; 32];
+    output[31] = if result { 1 } else { 0 };
+
+    Ok(output)
+}
+
+/// Helper function to perform the pairing check
+fn pairing_check(pairs: &[(blst_p1_affine, blst_p2_affine)]) -> bool {
+    if pairs.is_empty() {
+        return true;
+    }
+
+    // Compute the miller loop for the first pair
+    let (first_g1, first_g2) = &pairs[0];
```

Comment:
nit: you can write this as:

```suggestion
    let Some(((first_g1, first_g2), rest)) = pairs.split_first() else {
        return true;
    };
    
    // iter over rest without having to skip anything.
```

Honestly, the compiler will probably optimize them to the same thing, but it's easier to see that this version can't panic. But I'll leave this up to you in terms of readability (my tolerance for rust shenanigans is rather high).

Reply:
makes sense! ff006a13